ffi/examples/read-table/expression.h

```diff
@@ -0,0 +1,751 @@
+    lit->value.literal_field = val;                                                                \
+    return put_handle(data, lit, Literal);                                                         \
+  }                                                                                                \
+  _Static_assert(                                                                                  \
+    sizeof(c_type) <= sizeof(uintptr_t), "The provided type is not a valid simple scalar")
+#define DEFINE_VARIADIC(fun_name, enum_member)                                                     \
+  uintptr_t fun_name(void* data, uintptr_t len)                                                    \
+  {                                                                                                \
+    return visit_expr_variadic(data, len, enum_member);                                            \
+  }
+#define DEFINE_UNARY(fun_name, op)                                                                 \
+  uintptr_t fun_name(void* data, uintptr_t sub_expr)                                               \
+  {                                                                                                \
+    return visit_expr_unary(data, sub_expr, op);                                                   \
+  }
+enum OpType
+{
+  Add,
+  Sub,
+  Div,
+  Mul,
+  LT,
+  LE,
+  GT,
+  GE,
+  EQ,
+  NE,
+  Distinct,
+  In,
+  NotIn,
+};
+enum LitType
+{
+  Integer,
+  Long,
+  Short,
+  Byte,
+  Float,
+  Double,
+  String,
+  Boolean,
+  Timestamp,
+  TimestampNtz,
+  Date,
+  Binary,
+  Decimal,
+  Null,
+  Struct,
+  Array
+};
+enum ExpressionType
+{
+  BinOp,
+  Variadic,
+  Literal,
+  Unary,
+  Column
+};
+typedef struct
+{
+  void* ref;
+  enum ExpressionType type;
+} ExpressionRef;
+struct BinOp
+{
+  enum OpType op;
+  struct Literal* left;
+  struct Literal* right;
+};
+struct Null;
+
+enum VariadicType
+{
+  And,
+  Or,
+  StructConstructor,
+  ArrayData
+};
+enum UnaryType
+{
+  Not,
+  IsNull
+};
+struct Variadic
+{
+  enum VariadicType op;
+  size_t len;
+  size_t max_len;
+  ExpressionRef* expr_list;
+};
+struct Unary
+{
+  enum UnaryType type;
+  ExpressionRef sub_expr;
+};
+struct BinaryData
+{
+  uint8_t* buf;
+  uintptr_t len;
+};
+struct Decimal
+{
+  uint64_t value[2];
+  uint8_t precision;
+  uint8_t scale;
+};
+typedef struct
+{
+  size_t len;
+  ExpressionRef handles[100];
+} ExpressionBuilder;
+struct Struct
+{
+  KernelStringSlice* field_names;
+  ExpressionRef* expressions;
+  size_t len;
+  size_t max_len;
+};
+
+struct ArrayData
+{
+  size_t len;
+  size_t max_len;
+  ExpressionRef* expr_list;
+};
+
+struct Literal
+{
+  enum LitType type;
+  union LiteralValue
+  {
+    int32_t integer_data;
+    int64_t long_data;
+    int16_t short_data;
+    int8_t byte_data;
+    float float_data;
+    double double_data;
+    bool boolean_data;
+    struct KernelStringSlice string_data;
```

Comment:
You don't need this to be a `KernelStringSlice`, just make it a `char*`. Get rid of `copy_kernel_string()` and just allocate a c-string instead, like:
```c
// utility to turn a slice into a char*
char* allocate_string(const KernelStringSlice slice)
{
  return strndup(slice.ptr, slice.len);
}
```